devito/types/sparse.py

```diff
@@ -624,116 +835,6 @@ def _decomposition(self):
-        distributor = self.grid.distributor
-
-        # If not using MPI, don't waste time
-        if distributor.nprocs == 1:
-            return {self: data, self.coordinates: self.coordinates.data}
-
-        comm = distributor.comm
-        mpitype = MPI._typedict[np.dtype(self.dtype).char]
-
-        # Compute dist map only once
-        dmap = self._dist_datamap
-        mask = self._dist_scatter_mask(dmap=dmap)
-
-        # Pack sparse data values so that they can be sent out via an Alltoallv
-        data = data[mask]
-        data = np.ascontiguousarray(np.transpose(data, self._dist_reorder_mask))
-
-        # Send out the sparse point values
-        _, scount, sdisp, rshape, rcount, rdisp = self._dist_alltoall(dmap=dmap)
-        scattered = np.empty(shape=rshape, dtype=self.dtype)
-        comm.Alltoallv([data, scount, sdisp, mpitype],
-                       [scattered, rcount, rdisp, mpitype])
-        data = scattered
-
-        # Unpack data values so that they follow the expected storage layout
-        data = np.ascontiguousarray(np.transpose(data, self._dist_reorder_mask))
-
-        # Pack (reordered) coordinates so that they can be sent out via an Alltoallv
-        coords = self.coordinates.data._local[mask[self._sparse_position]]
-
-        # Send out the sparse point coordinates
-        _, scount, sdisp, rshape, rcount, rdisp = self._dist_subfunc_alltoall(dmap=dmap)
-        scattered = np.empty(shape=rshape, dtype=self.coordinates.dtype)
-        comm.Alltoallv([coords, scount, sdisp, mpitype],
-                       [scattered, rcount, rdisp, mpitype])
-        coords = scattered
-
-        # Translate global coordinates into local coordinates
-        coords = coords - np.array(self.grid.origin_offset, dtype=self.dtype)
-
-        return {self: data, self.coordinates: coords}
-
-    def _dist_gather(self, data, coords):
-        distributor = self.grid.distributor
-
-        # If not using MPI, don't waste time
-        if distributor.nprocs == 1:
-            return
-
-        comm = distributor.comm
-
-        # Compute dist map only once
-        dmap = self._dist_datamap
-        mask = self._dist_scatter_mask(dmap=dmap)
-
-        # Pack sparse data values so that they can be sent out via an Alltoallv
-        data = np.ascontiguousarray(np.transpose(data, self._dist_reorder_mask))
-        # Send back the sparse point values
-        sshape, scount, sdisp, _, rcount, rdisp = self._dist_alltoall(dmap=dmap)
-        gathered = np.empty(shape=sshape, dtype=self.dtype)
-        mpitype = MPI._typedict[np.dtype(self.dtype).char]
-        comm.Alltoallv([data, rcount, rdisp, mpitype],
-                       [gathered, scount, sdisp, mpitype])
-        # Unpack data values so that they follow the expected storage layout
-        gathered = np.ascontiguousarray(np.transpose(gathered, self._dist_reorder_mask))
-        self._data[mask] = gathered[:]
-
-        if coords is not None:
```

Comment:
I love it that subfunc and data scattering/gathering is now explicitly separated :)